cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/hbase/CoprocessorManager.java

```diff
@@ -0,0 +1,196 @@
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.jar.JarEntry;
+import java.util.jar.JarOutputStream;
+
+/**
+ * Manages HBase coprocessors for Tables and Queues.
+ */
+public class CoprocessorManager {
+  private static final Logger LOG = LoggerFactory.getLogger(CoprocessorManager.class);
+  protected final Location jarDir;
+  private final Map<Type, Set<Class<? extends Coprocessor>>> coprocessors;
+
+  /**
+   * Type of coprocessor.
+   */
+  public enum Type {
+    TABLE,
+    QUEUE,
+    MESSAGING
+  }
+
+  public CoprocessorManager(CConfiguration cConf, LocationFactory locationFactory, HBaseTableUtil tableUtil) {
+    this.jarDir = locationFactory.create(cConf.get(Constants.CFG_HDFS_LIB_DIR));
+    this.coprocessors = ImmutableMap.<Type, Set<Class<? extends Coprocessor>>>of(
+      Type.TABLE, ImmutableSet.of(tableUtil.getTransactionDataJanitorClassForVersion(),
+                                  tableUtil.getIncrementHandlerClassForVersion()),
+      Type.QUEUE, ImmutableSet.of(tableUtil.getQueueRegionObserverClassForVersion(),
+                                  tableUtil.getDequeueScanObserverClassForVersion()),
+      Type.MESSAGING, ImmutableSet.of(tableUtil.getMessageTableRegionObserverClassForVersion(),
+                                      tableUtil.getPayloadTableRegionObserverClassForVersion()));
+  }
+
+  /**
+   * Get the location of the specified type of coprocessor and ensure it exists.
+   * In distributed mode, the coprocessor jars are loaded onto hdfs by the CoprocessBuildTool,
+   * but in other modes it is still useful to create the jar on demand.
+   *
+   * @param type the type of coprocessor
+   * @return the location of the coprocessor
+   * @throws IOException if there was an issue accessing the location
+   * @throws IllegalStateException if the coprocessor jar does not exist
+   */
+  public synchronized Location ensureCoprocessorExists(Type type) throws IOException {
+
+    final Location targetPath = jarDir.append(String.format("%s-coprocessor-%s-%s.jar",
+                                                            type.name().toLowerCase(),
+                                                            ProjectInfo.getVersion(),
+                                                            HBaseVersion.get().toString()));
+    if (targetPath.exists()) {
+      return targetPath;
+    }
+
+    // ensure the jar directory exists
+    Locations.mkdirsIfNotExists(jarDir);
+
+    Set<Class<? extends Coprocessor>> classes = coprocessors.get(type);
+    StringBuilder buf = new StringBuilder();
+    for (Class<? extends Coprocessor> c : classes) {
+      buf.append(c.getName()).append(", ");
+    }
+    if (buf.length() == 0) {
+      return null;
+    }
+
+    LOG.debug("Creating jar file for coprocessor classes: " + buf.toString());
+    final Hasher hasher = Hashing.md5().newHasher();
+    final byte[] buffer = new byte[4 * 1024];
```

Comment:
better to move above two declarations close to where they are used. i.e. below the following for loop